File: packages/engine/src/scoring.rs
use crate::types::{Category, Dice, ScoringResult};

/// Count occurrences of each die value (1-6)
fn count_dice(dice: &Dice) -> [u8; 6] {
    let mut counts = [0u8; 6];
    for &d in dice {
        counts[(d - 1) as usize] += 1;
    }
    counts
}

/// Sum of all dice
fn sum_dice(dice: &Dice) -> u16 {
    dice.iter().map(|&d| d as u16).sum()
}

/// Score dice for a specific category
pub fn score(dice: &Dice, category: Category) -> ScoringResult {
    let counts = count_dice(dice);
    let total = sum_dice(dice);

    let (score, valid) = match category {
        // Upper section: sum of specific face value
        Category::Ones => {
            let s = counts[0] as u16;
            (s, true)
        }
        Category::Twos => {
            let s = counts[1] as u16 * 2;
            (s, true)
        }
        Category::Threes => {
            let s = counts[2] as u16 * 3;
            (s, true)
        }
        Category::Fours => {
            let s = counts[3] as u16 * 4;
            (s, true)
        }
        Category::Fives => {
            let s = counts[4] as u16 * 5;
            (s, true)
        }
        Category::Sixes => {
            let s = counts[5] as u16 * 6;
            (s, true)
        }

        // Three of a kind: at least 3 of same value
        Category::ThreeOfAKind => {
            let has_three = counts.iter().any(|&c| c >= 3);
            if has_three {
                (total, true)
            } else {
                (0, false)
            }
        }

        // Four of a kind: at least 4 of same value
        Category::FourOfAKind => {
            let has_four = counts.iter().any(|&c| c >= 4);
            if has_four {
                (total, true)
            } else {
                (0, false)
            }
        }

        // Full house: 3 of one + 2 of another
        Category::FullHouse => {
            let has_three = counts.contains(&3);
            let has_two = counts.contains(&2);
            if has_three && has_two {
                (25, true)
            } else {
                (0, false)
            }
        }

        // Small straight: 4 consecutive values
        Category::SmallStraight => {
            // Check for 1-2-3-4, 2-3-4-5, or 3-4-5-6
            let seq_1234 = counts[0] >= 1 && counts[1] >= 1 && counts[2] >= 1 && counts[3] >= 1;
            let seq_2345 = counts[1] >= 1 && counts[2] >= 1 && counts[3] >= 1 && counts[4] >= 1;
            let seq_3456 = counts[2] >= 1 && counts[3] >= 1 && counts[4] >= 1 && counts[5] >= 1;
            if seq_1234 || seq_2345 || seq_3456 {
                (30, true)
            } else {
                (0, false)
            }
        }

        // Large straight: 5 consecutive values
        Category::LargeStraight => {
            // Check for 1-2-3-4-5 or 2-3-4-5-6
            let has_large = (counts[0] == 1
                && counts[1] == 1
                && counts[2] == 1
                && counts[3] == 1
                && counts[4] == 1)
                || (counts[1] == 1
                    && counts[2] == 1
                    && counts[3] == 1
                    && counts[4] == 1
                    && counts[5] == 1);
            if has_large {
                (40, true)
            } else {
                (0, false)
            }
        }

        // Yahtzee: all 5 dice the same
        Category::Yahtzee => {
            let has_five = counts.contains(&5);
            if has_five {
                (50, true)
            } else {
                (0, false)
            }
        }

        // Chance: sum of all dice (always valid)
        Category::Chance => (total, true),
    };

    ScoringResult {
        category,
        score,
        valid,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_upper_section() {
        let dice: Dice = [1, 1, 2, 3, 4];
        assert_eq!(score(&dice, Category::Ones).score, 2);
        assert_eq!(score(&dice, Category::Twos).score, 2);
        assert_eq!(score(&dice, Category::Threes).score, 3);
        assert_eq!(score(&dice, Category::Fours).score, 4);
        assert_eq!(score(&dice, Category::Fives).score, 0);
        assert_eq!(score(&dice, Category::Sixes).score, 0);
    }

    #[test]
    fn test_three_of_a_kind() {
        let yes: Dice = [3, 3, 3, 2, 1];
        let no: Dice = [1, 2, 3, 4, 5];

        let result = score(&yes, Category::ThreeOfAKind);
        assert!(result.valid);
        assert_eq!(result.score, 12);

        let result = score(&no, Category::ThreeOfAKind);
        assert!(!result.valid);
        assert_eq!(result.score, 0);
    }

    #[test]
    fn test_four_of_a_kind() {
        let yes: Dice = [4, 4, 4, 4, 2];
        let no: Dice = [4, 4, 4, 2, 2];

        let result = score(&yes, Category::FourOfAKind);
        assert!(result.valid);
        assert_eq!(result.score, 18);

        let result = score(&no, Category::FourOfAKind);
        assert!(!result.valid);
    }

    #[test]
    fn test_full_house() {
        let yes: Dice = [3, 3, 3, 2, 2];
        let no: Dice = [3, 3, 3, 3, 2];

        let result = score(&yes, Category::FullHouse);
        assert!(result.valid);
        assert_eq!(result.score, 25);

        let result = score(&no, Category::FullHouse);
        assert!(!result.valid);
    }

    #[test]
    fn test_small_straight() {
        let yes1: Dice = [1, 2, 3, 4, 6];
        let yes2: Dice = [2, 3, 4, 5, 5];
        let yes3: Dice = [1, 3, 4, 5, 6];
        let no: Dice = [1, 2, 3, 5, 6];

        assert!(score(&yes1, Category::SmallStraight).valid);
        assert!(score(&yes2, Category::SmallStraight).valid);
        assert!(score(&yes3, Category::SmallStraight).valid);
        assert!(!score(&no, Category::SmallStraight).valid);
    }

    #[test]
    fn test_large_straight() {
        let yes1: Dice = [1, 2, 3, 4, 5];
        let yes2: Dice = [2, 3, 4, 5, 6];
        let no: Dice = [1, 2, 3, 4, 6];

        let result = score(&yes1, Category::LargeStraight);
        assert!(result.valid);
        assert_eq!(result.score, 40);

        assert!(score(&yes2, Category::LargeStraight).valid);
        assert!(!score(&no, Category::LargeStraight).valid);
    }

    #[test]
    fn test_yahtzee() {
        let yes: Dice = [5, 5, 5, 5, 5];
        let no: Dice = [5, 5, 5, 5, 4];

        let result = score(&yes, Category::Yahtzee);
        assert!(result.valid);
        assert_eq!(result.score, 50);

        assert!(!score(&no, Category::Yahtzee).valid);
    }

    #[test]
    fn test_chance() {
        let dice: Dice = [1, 2, 3, 4, 5];
        let result = score(&dice, Category::Chance);
        assert!(result.valid);
        assert_eq!(result.score, 15);
    }
}

File: packages/cloudflare-do/src/game/scoring.ts
/**
 * Score Calculation
 *
 * Server-authoritative scoring for all Dicee categories.
 * Uses crypto.getRandomValues for dice generation.
 *
 * Migrated from packages/partykit - no functional changes.
 */

import type { Category, DiceArray, Scorecard } from './types';
import { DICEE_BONUS_VALUE, UPPER_BONUS_THRESHOLD, UPPER_BONUS_VALUE } from './types';

// =============================================================================
// Dice Generation (Server-Authoritative)
// =============================================================================

/**
 * Generate random dice values using cryptographic randomness
 */
export function generateDice(count: 5): DiceArray;
export function generateDice(count: number): number[];
export function generateDice(count: number): number[] {
	const buffer = new Uint8Array(count);
	crypto.getRandomValues(buffer);
	return Array.from(buffer).map((byte) => (byte % 6) + 1);
}

/**
 * Roll dice with kept mask - only re-roll dice not kept
 */
export function rollWithKept(
	currentDice: DiceArray | null,
	kept: [boolean, boolean, boolean, boolean, boolean],
): DiceArray {
	// Generate new random values
	const newValues = generateDice(5);

	// If no current dice, return all new
	if (!currentDice) {
		return newValues as DiceArray;
	}

	// Apply kept mask
	return [
		kept[0] ? currentDice[0] : newValues[0],
		kept[1] ? currentDice[1] : newValues[1],
		kept[2] ? currentDice[2] : newValues[2],
		kept[3] ? currentDice[3] : newValues[3],
		kept[4] ? currentDice[4] : newValues[4],
	];
}

// =============================================================================
// Score Calculation Helpers
// =============================================================================

/**
 * Count occurrences of each dice value
 */
function countValues(dice: DiceArray): Map<number, number> {
	const counts = new Map<number, number>();
	for (const value of dice) {
		counts.set(value, (counts.get(value) ?? 0) + 1);
	}
	return counts;
}

/**
 * Sum all dice
 */
function sumAll(dice: DiceArray): number {
	return dice.reduce((sum, v) => sum + v, 0);
}

/**
 * Sum dice matching a specific value
 */
function sumMatching(dice: DiceArray, value: number): number {
	return dice.filter((v) => v === value).reduce((sum, v) => sum + v, 0);
}

/**
 * Check if dice contain n of a kind
 */
function hasOfAKind(dice: DiceArray, n: number): boolean {
	const counts = countValues(dice);
	for (const count of counts.values()) {
		if (count >= n) return true;
	}
	return false;
}

/**
 * Check for full house (3 of one, 2 of another)
 */
function isFullHouse(dice: DiceArray): boolean {
	const counts = countValues(dice);
	const values = Array.from(counts.values()).sort((a, b) => a - b);
	return values.length === 2 && values[0] === 2 && values[1] === 3;
}

/**
 * Check for small straight (4 consecutive)
 */
function isSmallStraight(dice: DiceArray): boolean {
	const unique = new Set(dice);
	// Possible small straights: 1234, 2345, 3456
	const patterns = [
		[1, 2, 3, 4],
		[2, 3, 4, 5],
		[3, 4, 5, 6],
	];
	return patterns.some((pattern) => pattern.every((v) => unique.has(v)));
}

/**
 * Check for large straight (5 consecutive)
 */
function isLargeStraight(dice: DiceArray): boolean {
	const sorted = [...dice].sort((a, b) => a - b);
	// Must be either 1-2-3-4-5 or 2-3-4-5-6
	return (
		(sorted[0] === 1 && sorted[1] === 2 && sorted[2] === 3 && sorted[3] === 4 && sorted[4] === 5) ||
		(sorted[0] === 2 && sorted[1] === 3 && sorted[2] === 4 && sorted[3] === 5 && sorted[4] === 6)
	);
}

/**
 * Check for Dicee (5 of a kind)
 */
function isDicee(dice: DiceArray): boolean {
	return new Set(dice).size === 1;
}

// =============================================================================
// Category Score Calculation
// =============================================================================

/**
 * Calculate score for a specific category
 */
export function calculateCategoryScore(dice: DiceArray, category: Category): number {
	switch (category) {
		// Upper section - sum of matching dice
		case 'ones':
			return sumMatching(dice, 1);
		case 'twos':
			return sumMatching(dice, 2);
		case 'threes':
			return sumMatching(dice, 3);
		case 'fours':
			return sumMatching(dice, 4);
		case 'fives':
			return sumMatching(dice, 5);
		case 'sixes':
			return sumMatching(dice, 6);

		// Lower section
		case 'threeOfAKind':
			return hasOfAKind(dice, 3) ? sumAll(dice) : 0;
		case 'fourOfAKind':
			return hasOfAKind(dice, 4) ? sumAll(dice) : 0;
		case 'fullHouse':
			return isFullHouse(dice) ? 25 : 0;
		case 'smallStraight':
			return isSmallStraight(dice) ? 30 : 0;
		case 'largeStraight':
			return isLargeStraight(dice) ? 40 : 0;
		case 'dicee':
			return isDicee(dice) ? 50 : 0;
		case 'chance':
			return sumAll(dice);
	}
}

/**
 * Calculate all potential scores for current dice
 */
export function calculateAllPotentialScores(dice: DiceArray): Record<Category, number> {
	return {
		ones: calculateCategoryScore(dice, 'ones'),
		twos: calculateCategoryScore(dice, 'twos'),
		threes: calculateCategoryScore(dice, 'threes'),
		fours: calculateCategoryScore(dice, 'fours'),
		fives: calculateCategoryScore(dice, 'fives'),
		sixes: calculateCategoryScore(dice, 'sixes'),
		threeOfAKind: calculateCategoryScore(dice, 'threeOfAKind'),
		fourOfAKind: calculateCategoryScore(dice, 'fourOfAKind'),
		fullHouse: calculateCategoryScore(dice, 'fullHouse'),
		smallStraight: calculateCategoryScore(dice, 'smallStraight'),
		largeStraight: calculateCategoryScore(dice, 'largeStraight'),
		dicee: calculateCategoryScore(dice, 'dicee'),
		chance: calculateCategoryScore(dice, 'chance'),
	};
}

// =============================================================================
// Scorecard Updates
// =============================================================================

/**
 * Apply a score to a category and update bonuses
 * Returns updated scorecard and whether it was a Dicee bonus
 */
export function applyScore(
	scorecard: Scorecard,
	category: Category,
	dice: DiceArray,
): { scorecard: Scorecard; score: number; isDiceeBonus: boolean } {
	const score = calculateCategoryScore(dice, category);
	let isDiceeBonus = false;

	// Create new scorecard (immutable update)
	const newScorecard: Scorecard = { ...scorecard };

	// Check for Dicee bonus
	// If scoring Dicee and already have a Dicee scored (non-zero), add bonus
	if (isDicee(dice) && scorecard.dicee !== null && scorecard.dicee > 0) {
		newScorecard.diceeBonus = scorecard.diceeBonus + DICEE_BONUS_VALUE;
		isDiceeBonus = true;
	}

	// Apply the score
	newScorecard[category] = score;

	// Recalculate upper bonus
	const upperSum =
		(newScorecard.ones ?? 0) +
		(newScorecard.twos ?? 0) +
		(newScorecard.threes ?? 0) +
		(newScorecard.fours ?? 0) +
		(newScorecard.fives ?? 0) +
		(newScorecard.sixes ?? 0);

	if (upperSum >= UPPER_BONUS_THRESHOLD && newScorecard.upperBonus === 0) {
		newScorecard.upperBonus = UPPER_BONUS_VALUE;
	}

	return { scorecard: newScorecard, score, isDiceeBonus };
}

/**
 * Calculate total score from a scorecard
 */
export function calculateTotal(scorecard: Scorecard): number {
	let total = 0;

	// Upper section
	total += scorecard.ones ?? 0;
	total += scorecard.twos ?? 0;
	total += scorecard.threes ?? 0;
	total += scorecard.fours ?? 0;
	total += scorecard.fives ?? 0;
	total += scorecard.sixes ?? 0;
	total += scorecard.upperBonus;

	// Lower section
	total += scorecard.threeOfAKind ?? 0;
	total += scorecard.fourOfAKind ?? 0;
	total += scorecard.fullHouse ?? 0;
	total += scorecard.smallStraight ?? 0;
	total += scorecard.largeStraight ?? 0;
	total += scorecard.dicee ?? 0;
	total += scorecard.chance ?? 0;
	total += scorecard.diceeBonus;

	return total;
}
